### PR TITLE
Speedy Block Brakes

### PR DESCRIPTION
**Pull request by spacek531**
This is a continuation of #13838 since it was closed and unable to reopen.

This PR adds speedy block brakes, a long-requested feature. Block brakes now have a speed field which allows users to set the maximum speed a train may traverse the block brake. Vehicles stopped by the block brake will resume motion at the original speed when the block brake opens.

To compliment the speedy block brakes, regular brakes have new behavior. When an uninterrupted line of brakes precede a block brake, the brakes will disable themselves while these conditions are met:
- The block brake is open
- The brake's speed is lower than the block brake's speed

In addition to these two features, the following supporting modifications are made:
- Tile inspector changes checkbox text "block brake closed" to "brake closed" when selecting a non-block brake piece
- Track designs and older parks sets all block brake speed to default, which is identical to legacy behavior
- Special case added to track design export to overwrite speed field with rotation field when saving block brakes on multidim coasters
- Removed arbitrary limitation of TrackPlaceAction that prevents building track element with speed and seat rotation

### CODE REVIEW

**Comment by spacek531** [comment]
Will require replays update.

**Comment by karst** [comment]
Seems like you have to poke the compiler again. Seems to do this often lately. It builds just fine on my machine.

**Comment by karst** [comment]
image
Is it intentional you built this on top of 0.3.5.1?

**Comment by duncanspumpkin** [comment]
As this will break existing track designs i can't see how we can add it like this.

**Comment by karst** [comment]
I don't see how it would break existing track designs if they all default to what it'd be before this change. I could see people save the design without being able to complete when coming to a full stop though, idk if there was thought about that.

**Comment by spacek531** [comment]
> As this will break existing track designs i can't see how we can add it like this.

Can you post which designs it breaks?

**Comment by duncanspumpkin** [comment]
The fact that the replays are invalidated tells me that it changes the logic of vehicles and therefore will break designs

**Comment by spacek531** [comment]
> Seems like you have to poke the compiler again. Seems to do this often lately. It builds just fine on my machine.

It compiled successfully, the errors are from BPB 1000 ticks

**Comment by spacek531** [comment]
> The fact that the replays are invalidated tells me that it changes the logic of vehicles and therefore will break designs

Replays are invalidated because it is loading a value where nine was loaded before. The import code properly handles Legacy block brakes and updates them to use the exact same speed that they had before.

**Comment by duncanspumpkin** [comment]
> > The fact that the replays are invalidated tells me that it changes the logic of vehicles and therefore will break designs
> 
> Replays are invalidated because it is loading a value where none was loaded before. The import code properly handles Legacy block brakes and updates them to use the exact same speed that they had before.

How do you achieve the legacy behaviour of block brakes there is no way to select that. Therefore it will change the behaviour of the vehicles and break tracks

**Comment by spacek531** [comment]
> > > The fact that the replays are invalidated tells me that it changes the logic of vehicles and therefore will break designs
> > 
> > 
> > Replays are invalidated because it is loading a value where none was loaded before. The import code properly handles Legacy block brakes and updates them to use the exact same speed that they had before.
> 
> How do you achieve the legacy behaviour of block brakes there is no way to select that. Therefore it will change the behaviour of the vehicles and break tracks

Block brakes set to 4 mph use the legacy speed. Track designs and parks import all block brakes at 4 mph. Therefore legacy behavior is maintained.

**Comment by spacek531** [comment]
BlockBrakeSetLinkedBrakesClosed has a guard clause for trackType so it exits early if the track it is called for is not a block brake. I also have a check for the track type *outside* the function that serves the same purpose. Are the redundant checks bad? Is it preferred to have the guard clause or to check before calling the function?

**Comment by X123M3-256** [comment]
Any update on this?

**Comment by spacek531** [comment]
[replay_desync_84964.txt](https://github.com/OpenRCT2/OpenRCT2/files/10231622/replay_desync_84964.txt)

Here is the desync file from the current bpb1000ticks. The desync genotype is that `brake_speed` is changed from 0 to 2 as expected. The desync phenotype is nothing - there are no vehicle position desyncs which indicates the behavior is perfectly legacy compatible.

File: data/language/en-GB.txt [hunk 1]
```diff
@@ -3639,6 +3639,10 @@ STR_6533    :{WINDOW_COLOUR_2}Excitement Factor: {BLACK}-{COMMA16}%
 STR_6534    :{WINDOW_COLOUR_2}Intensity Factor: {BLACK}-{COMMA16}%
 STR_6535    :{WINDOW_COLOUR_2}Nausea Factor: {BLACK}-{COMMA16}%
 STR_6536    :This park was saved in a later version of OpenRCT2. The park was saved in v{INT32}, you are currently on v{INT32}.
+STR_6537    :Brake Closed
+STR_6538    :Block brake speed
+STR_6539    :Set speed limit for block brakes. In block section mode, adjacent brakes with a slower speed are linked to the block brake
+STR_6540    :Block brakes will be saved at default speed
```
Comment (spacek531) [inline data/language/en-GB.txt:3645]: Currently unused, waiting for #18771 merged to be used.

**Comment by spacek531** [comment]
Note: bump network version

File: src/openrct2/rct12/RCT12.cpp [hunk 2]
```diff
@@ -477,7 +478,7 @@ bool RCT12PathElement::IsBlockedByVehicle() const
     return (flags & RCT12_TILE_ELEMENT_FLAG_BLOCKED_BY_VEHICLE) != 0;
 }
 
-bool RCT12TrackElement::BlockBrakeClosed() const
+bool RCT12TrackElement::GetBrakeClosed() const
```
Comment (duncanspumpkin) [inline src/openrct2/rct12/RCT12.cpp:481]: This should not have been changed. Its still the block brake closed.

File: src/openrct2/ride/Vehicle.cpp [hunk 3]
```diff
@@ -6067,9 +6067,10 @@ void Vehicle::ApplyNonStopBlockBrake()
             velocity = BLOCK_BRAKE_BASE_SPEED;
             acceleration = 0;
         }
-        else
+        // the 868 offset is to make the min speed equal to
+        // BLOCK_BRAKE_BASE_SPEED to keep existing behaviour
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Vehicle.cpp:6071]: Why not use the `BLOCK_BRAKE_BASE_SPEED` constant then?

File: src/openrct2/rct12/RCT12.cpp (hunk 2 above)
Comment (spacek531) [reply src/openrct2/rct12/RCT12.cpp:481]: Regular brakes have a closed state in this PR.

File: src/openrct2/ride/Vehicle.cpp (hunk 3 above)
Comment (spacek531) [reply src/openrct2/ride/Vehicle.cpp]: It would be messy if I used `BLOCK_BRAKE_BASE_SPEED` in the equality:
```cpp
else if (velocity > (brake_speed << 16) + (BLOCK_BRAKE_BASE_SPEED - (RCT2DefaultBlockBrakeSpeed << 16))
```

I made a constant so it's no longer a magic number.

File: src/openrct2/rct12/RCT12.cpp (hunk 2 above)
Comment (spacek531) [reply src/openrct2/rct12/RCT12.cpp]: I missed the RCT12 part. This is now fixed.

File: src/openrct2/ride/Ride.cpp [hunk 4]
```diff
@@ -3026,19 +3026,99 @@ static void RideOpenBlockBrakes(CoordsXYE* startElement)
...
             case TrackElemType::DiagUp60ToFlat:
-            case TrackElemType::BlockBrakes:
-                currentElement.element->AsTrack()->SetBlockBrakeClosed(false);
+                currentElement.element->AsTrack()->SetBrakeClosed(false);
                 break;
         }
     } while (track_block_get_next(&currentElement, &currentElement, nullptr, nullptr)
              && currentElement.element != startElement->element);
 }
+void BrakeLinkToBlockBrake(const CoordsXYZ& vehicleTrackLocation, TileElement* tileElement)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:3046]: pass a trackelement instead of tile element this function only works with tracks.

Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:3046]: pass a trackelement instead of tile element this function only works with tracks.

File: src/openrct2/ride/Ride.cpp [hunk 5]
```diff
@@ -3026,19 +3026,99 @@ static void RideOpenBlockBrakes(CoordsXYE* startElement)
...
+        }
+        else
+        {
+            brake->SetBrakeClosed(true);
+            break;
+        }
+    } while (track_block_get_next(&output, &output, &outputZ, nullptr));
+}
+
+void BlockBrakeSetLinkedBrakesClosed(const CoordsXYZ& vehicleTrackLocation, TileElement* tileElement, bool isClosed)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:3073]: trackelement

File: src/openrct2/ride/Ride.cpp [hunk 6]
```diff
@@ -3551,14 +3631,19 @@ ResultWithMessage Ride::CreateVehicles(const CoordsXYE& element, bool isApplying
  * preceding that block.
  *  rct2: 0x006DDF9C
  */
-void Ride::MoveTrainsToBlockBrakes(TrackElement* firstBlock)
+void Ride::MoveTrainsToBlockBrakes(CoordsXYE* currentElement)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:3634]: id prefer this was a coordsxy and a trackelement. Coordsxye is a bit of a rubbish structure and we should limit its use.

File: src/openrct2/ride/Vehicle.cpp [hunk 7]
```diff
@@ -7596,7 +7601,9 @@ bool Vehicle::UpdateTrackMotionForwards(CarEntry* carEntry, Ride* curRide, rct_r
             auto brakeSpeed = brake_speed << 16;
             if (brakeSpeed < _vehicleVelocityF64E08)
             {
-                acceleration = -_vehicleVelocityF64E08 * 16;
+                auto trackElement = MapGetTrackElementAtOfTypeSeq(TrackLocation, trackType, 0);
+                if ((trackElement != nullptr && trackElement->AsTrack()->GetBrakeClosed()) || trackElement == nullptr)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Vehicle.cpp:7605]: this is odd? Why care about the nullptr case?

File: src/openrct2/ride/Ride.cpp (hunk 4 above)
Comment (spacek531) [reply src/openrct2/ride/Ride.cpp:3046]: CoordsXY and TrackElement? I thought TrackElement does not have position data

File: src/openrct2/actions/TrackPlaceAction.cpp [hunk 8]
```diff
@@ -625,12 +625,17 @@ GameActions::Result TrackPlaceAction::Execute() const
...
+                break;
         }
+
         if (TrackTypeHasSpeedSetting(_trackType))
         {
             trackElement->SetBrakeBoosterSpeed(_brakeSpeed);
         }
-        else if (ride->GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_HAS_LANDSCAPE_DOORS))
+
+        if (ride->GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_HAS_LANDSCAPE_DOORS))
```
Comment (duncanspumpkin) [inline src/openrct2/actions/TrackPlaceAction.cpp:638]: why this change? The logic is a bit silly nowadays since the seat and speed are split just want to know if this was to fix/address an issue.

File: src/openrct2/ride/Vehicle.cpp (hunk 7 above)
Comment (spacek531) [reply src/openrct2/ride/Vehicle.cpp:7605]: If the vehicle believes it is on a brake piece but the track element cannot be found to verify the brake is closed, assume the brake is closed

File: src/openrct2/ride/Ride.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/openrct2/ride/Ride.cpp:3046]: yes `const CoordsXYZ& vehicleTrackLocation, TrackElement&`

File: src/openrct2/actions/TrackPlaceAction.cpp (hunk 8 above)
Comment (spacek531) [reply src/openrct2/actions/TrackPlaceAction.cpp:638]: The original only allows setting seat rotation if speed is not set - this PR supports speed and seat rotation on block brakes because td6 encodes seat rotation not block brake speed.

**Comment by Gymnasiast** [comment]
Why does this need changes to regular brakes? I would expect the changes to be confined to just the block brakes themselves.

**Comment by spacek531** [comment]
It is desirable to affect brakes so that they can gently slow down a train when the next block is occupied, but *not* slow down the train if the next block is free, to create the "flying" MCBR, as seen here:
https://youtu.be/gmKq6mEsrV4?t=49
Diamondback at Kings Island, in the first clip, demonstrates the desired flying performance.
Dive Coaster at Chimelong Paradise, demonstrates how MCBRs trim speed - it's a gradual reduction in speed over the length of the brake run.

Were I to build a flying MCBR with only block brake behaviour changed, I would be required to set the brake speeds to above the block brake speed. This works fine if the MCBR is always open, but if the ride stacks for any reason, then the train will forcefully jerk to a halt instead of the gradual reduction in speed. The solution was to selectively disable brakes.

image

**Comment by karst** [comment]
I'd also like to mention Steel Vengeance's mid-course. https://youtu.be/RTXTbzctl0c?t=85 As you can see the brakes didn't initiate at all here since the next block is clear. So changing the behaviour of regular brakes with it definitely makes sense.

**Comment by spacek531** [comment]
I have rebased with what I hope is a more logical order to the commits.

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 9]
```diff
@@ -1809,12 +1816,19 @@ class RideConstructionWindow final : public Window
...
-        bool brakesSelected = _selectedTrackType == TrackElemType::Brakes
-            || _currentTrackCurve == (RideConstructionSpecialPieceSelected | TrackElemType::Brakes);
+        bool trackHasSpeedSetting = TrackTypeHasSpeedSetting(_selectedTrackType)
+            || TrackTypeHasSpeedSetting(_currentTrackCurve & ~RideConstructionSpecialPieceSelected);
         bool boosterTrackSelected = _selectedTrackType == TrackElemType::Booster
             || _currentTrackCurve == (RideConstructionSpecialPieceSelected | TrackElemType::Booster);
 
-        if (!brakesSelected && !boosterTrackSelected)
+        // Only necessary because TD6 writes speed and seat rotation to the same bits. Remove for new track design format.
+        bool trackHasSpeedAndSeatRotation = _selectedTrackType == TrackElemType::BlockBrakes
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/RideConstruction.cpp:1825]: This doesn’t actually check for seat rotation in any way, so why is it called this?

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 10]
```diff
@@ -1854,60 +1868,73 @@ class RideConstructionWindow final : public Window
...
-            widgets[WIDX_BANK_RIGHT].right = 83;
-            widgets[WIDX_BANK_RIGHT].top = 139;
-            widgets[WIDX_BANK_RIGHT].bottom = 148;
-            hold_down_widgets |= (1uLL << WIDX_BANK_STRAIGHT) | (1uLL << WIDX_BANK_RIGHT);
-        }
-
-        widgets[WIDX_BANKING_GROUPBOX].right = 162;
+            widgets[WIDX_SPEED_SETTING_SPINNER].text = STR_RIDE_CONSTRUCTION_BRAKE_SPEED_VELOCITY;
+
+            widgets[WIDX_SPEED_SETTING_SPINNER].type = WindowWidgetType::Spinner;
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/RideConstruction.cpp:1889]: Why are all these changes needed?

File: src/openrct2/park/ParkFile.cpp [hunk 11]
```diff
@@ -1071,6 +1075,31 @@ namespace OpenRCT2
             }
         }
 
+        void UpdateBrakesClosedState()
+        {
+            for (int32_t y = 0; y < MAXIMUM_MAP_SIZE_TECHNICAL; y++)
```
Comment (Gymnasiast) [inline src/openrct2/park/ParkFile.cpp:1080]: This doesn’t have to search the whole 1001 × 1001 tiles - use the actual map size here.

File: src/openrct2/ride/Station.cpp [hunk 12]
```diff
@@ -61,7 +61,7 @@ static void ride_update_station_blocksection(Ride* ride, StationIndex stationInd
     auto& station = ride->GetStation(stationIndex);
 
     if ((ride->status == RideStatus::Closed && ride->num_riders == 0)
-        || (tileElement != nullptr && tileElement->AsTrack()->BlockBrakeClosed()))
+        || (tileElement != nullptr && tileElement->AsTrack()->GetBrakeClosed()))
```
Comment (Gymnasiast) [inline src/openrct2/ride/Station.cpp:64]: Rename this method to `IsBrakeClosed()`

File: src/openrct2/ride/Track.h [hunk 13]
```diff
@@ -18,7 +18,9 @@
 
 constexpr const uint32_t RideConstructionSpecialPieceSelected = 0x10000;
 
-constexpr const int32_t BLOCK_BRAKE_BASE_SPEED = 0x20364;
+constexpr const uint8_t RCT2DefaultBlockBrakeSpeed = 2;
+constexpr const int32_t BlockBrakeBaseSpeed = 0x20364;
+constexpr const int32_t BlockBrakeSpeedOffset = BlockBrakeBaseSpeed - (RCT2DefaultBlockBrakeSpeed << 16);
```
Comment (Gymnasiast) [inline src/openrct2/ride/Track.h:23]: Could you explain `BlockBrakeSpeedOffset`?

Comment (spacek531) [reply src/openrct2/ride/Track.h:23]: Chris sawyer chose the block brake speed threshold to be (2 << 16) + 868 instead of say (2 << 16)

File: src/openrct2-ui/windows/RideConstruction.cpp (hunk 10 above)
Comment (spacek531) [reply src/openrct2-ui/windows/RideConstruction.cpp:1889]: If you are asking why the UI was changed to accomodate speed and seat rotation at the same time, it's because block brakes support both.

File: src/openrct2-ui/windows/RideConstruction.cpp (hunk 9 above)
Comment (spacek531) [reply src/openrct2-ui/windows/RideConstruction.cpp:1825]: Those are all the track elements that have a speed setting and support rotation. Technically it should be block brakes and any track element above HighestAlias since they cannot be written to td6.

**Comment by Gymnasiast** [comment]
Could you also create a separate PR for the changes to the Construction window?

**Comment by Legotown55** [comment]
Will this be murged this weekend

**Comment by Gymnasiast** [comment]
> Will this be murged this weekend

If you don’t stop asking this on every PR, I’m going to ban you.

**Comment by duncanspumpkin** [comment]
[replay_desync_85014.txt](https://github.com/OpenRCT2/OpenRCT2/files/10421136/replay_desync_85014.txt)
[replay_desync_84014.txt](https://github.com/OpenRCT2/OpenRCT2/files/10421137/replay_desync_84014.txt)
For reference these are the desync reports. Nothing of concern in them.

File: src/openrct2/ride/Ride.cpp [hunk 14]
```diff
@@ -3026,19 +3026,98 @@ static void RideOpenBlockBrakes(const CoordsXYE& startElement)
...
+{
+    CoordsXYE output = CoordsXYE(vehicleTrackLocation.x, vehicleTrackLocation.y, reinterpret_cast<TileElement*>(&brake));
+    int32_t outputZ = vehicleTrackLocation.z;
+    do
+    {
+        if (output.element->AsTrack()->GetTrackType() == TrackElemType::BlockBrakes)
+        {
+            brake.SetBrakeClosed(
+                (brake.GetBrakeBoosterSpeed() >= output.element->AsTrack()->GetBrakeBoosterSpeed())
+                || (output.element->AsTrack()->IsBrakeClosed()));
```
Comment (Gymnasiast) [inline src/openrct2/ride/Ride.cpp:3056]: This can be much clearer. 

```
const auto blockBrakeSpeedIsLower = output.element->AsTrack()->GetBrakeBoosterSpeed() < brake.GetBrakeBoosterSpeed();
const auto blockBrakeIsClosed = output.element->AsTrack()->IsBrakeClosed();
const auto closeBrake = blockBrakeSpeedIsLower || blockBrakeIsClosed;

File: src/openrct2/ride/Ride.cpp [hunk 15]
```diff
@@ -3026,19 +3026,98 @@ static void RideOpenBlockBrakes(const CoordsXYE& startElement)
...
+        }
+        else if (output.element->AsTrack()->GetTrackType() == TrackElemType::Brakes)
+        {
+            continue;
+        }
+        else
+        {
+            brake.SetBrakeClosed(true);
+            break;
+        }
```
Comment (Gymnasiast) [inline src/openrct2/ride/Ride.cpp:3067]: If I understand the algorithm correctly, it goes like this:
1. Pick a brake.
2. Search forward
2.1. If we found a block brake, check if we need the close.
2.2. If we found another brake, keep searching until we find a block brake
2.3. If we found another piece, there is no block brake, so always close the brake.
4. Repeat step 1 with another brake until all are done.

Is this correct?

File: src/openrct2/ride/Ride.cpp [hunk 16]
```diff
@@ -3026,19 +3026,98 @@ static void RideOpenBlockBrakes(const CoordsXYE& startElement)
...
+        if (trackBeginEnd.begin_x == vehicleTrackLocation.x && trackBeginEnd.begin_y == vehicleTrackLocation.y
+            && tileElement == trackBeginEnd.begin_element)
+        {
+            return;
+        }
+
+        location.x = trackBeginEnd.end_x;
+        location.y = trackBeginEnd.end_y;
+        location.z = trackBeginEnd.begin_z;
```
Comment (Gymnasiast) [inline src/openrct2/ride/Ride.cpp:3095]: Is this correct? You use end\_x, end\_y and _begin_\_z here.

**Comment by Gymnasiast** [comment]
I tested it. I noticed the following scenario, which I’m not sure is correct.

I have a set of regular brakes, all set to 28 km/h, followed by a 35 km/h block brake. The block brake is open. A train arrives running at > 40 km/h, and does not get slowed down at all. This doesn’t seem correct to me. (I know why the code behaves like this, but this doesn’t seem how brakes should behave to me.)

**Comment by karst** [comment]
Hmm.. I don't see why it shouldn't function like that. Say if I'd want the train to slow down gradually if the block is closed then I'd want to have that kind of brake speed, but if it's clear having it go through at a maximum speed that's higher than that brake speed might be more desirable. I kinda would always want the brakes before the block be slower than the exit speed if the blockzone is clear due to this graduality.

File: src/openrct2/ride/Ride.cpp (hunk 15 above)
Comment (spacek531) [reply src/openrct2/ride/Ride.cpp:3057]: 1 and 2 yes. This function is only called from TrackSetSpeedAction so it will only affect one TrackElement at a time - ride simulation uses BlockBrakeSetLinkedBrakesClosed.

Note to self: add calls to this and BlockBrakeSetLinkedBrakesClosed in TrackPlaceAction,

**Comment by spacek531** [comment]
> I tested it. I noticed the following scenario, which I’m not sure is correct.
> 
> I have a set of regular brakes, all set to 28 km/h, followed by a 35 km/h block brake. The block brake is open. A train arrives running at > 40 km/h, and does not get slowed down at all. This doesn’t seem correct to me. (I know why the code behaves like this, but this doesn’t seem how brakes should behave to me.) IMO, the brakes should at least bring down the speed to 35 km/h (the passing speed of the block brake).

Did the block brake slow the train to 35 km/h?

**Comment by karst** [comment]
I have just retested the exact scenario, but I cannot seem to replicate any error. It does go down to 35 kmh.
https://i.gyazo.com/b11ecc8199887286382e6e2fc2664045.gif

**Comment by karst** [comment]
I was thinking, tossing it into discussion. Should you raise the minimum version to 17 as well? Wouldn't want to load up a park in an older version and suddenly have coasters no longer function correctly.

**Comment by spacek531** [comment]
There's no reason to do that if users design coasters so they still work at 4 mph brakes. This is the same as vanilla and represents a restart after a ride stoppage.

**Comment by Legotown55** [comment]
how do i get to test speedy block breaks

**Comment by karst** [comment]
> how do i get to test speedy block breaks

Usually under dependency checks there's artifacrs you can download. If not you have to load this pr into your OpenRCT2 fork yourself and build it.

**Comment by Legotown55** [comment]
how do i load a PR into open rct 2 myself

**Comment by spacek531** [comment]
For future reference:

test park with 10 wooden twister coasters with 12-car trains with 20 tile-long brakeruns with opengl render: 36-37 fps

**Comment by ZehMatt** [comment]
I'm a bit curious if it wouldn't make perhaps more sense to have a new track piece called "Speed Controller" which adds all those special things it can break or accelerate. I'm not sure if brakes should ever accelerate something, by definition brakes slow things down and that's about it, my question would be: any particular reason we can't just have a new element vs hacking new functionality into an existing element?

**Comment by karst** [comment]
> I'm a bit curious if it wouldn't make perhaps more sense to have a new track piece called "Speed Controller" which adds all those special things it can break or accelerate. I'm not sure if brakes should ever accelerate something, by definition brakes slow things down and that's about it, my question would be: any particular reason we can't just have a new element vs hacking new functionality into an existing element?

@X123M3-256 was working on a new implementation of Boosters that would give such abilities. I do not know how far he is with that or planning on finishing it, but it looks like a very fun and quite logical implementation. I believe he was mostly stuck on UI, though don't quote me on that. 

In this case it doesn't actually speed up trains to the block speed, if it's closed it still just departs at 6kmh, but if the block is open it will rush over it at that speed.

**Comment by ZehMatt** [comment]
> > I'm a bit curious if it wouldn't make perhaps more sense to have a new track piece called "Speed Controller" which adds all those special things it can break or accelerate. I'm not sure if brakes should ever accelerate something, by definition brakes slow things down and that's about it, my question would be: any particular reason we can't just have a new element vs hacking new functionality into an existing element?
> 
> @X123M3-256 was working on a new implementation of Boosters that would give such abilities. I do not know how far he is with that or planning on finishing it, but it looks like a very fun and quite logical implementation. I believe he was mostly stuck on UI, though don't quote me on that.
> 
> In this case it doesn't actually speed up trains to the block speed (unless ofcourse you would put a booster there), if it's closed it still just departs at 6kmh, but if the block is open it will rush over it at that speed.

That's why I think something like "Speed Controller" makes more sense where you can just set the right properties and you get pretty much what you want, could even have stupid things attributes like apply instantaneously and what not, unfortunately this approach is currently not compatible with the track design exporter, but I personally feel like this shouldn't be a stopper, we can just temporarily disallow exporting certain rides that contain unsupported track elements, but thats a topic on its own, its probably also best to to tackle that for once.

**Comment by karst** [comment]
> we can just temporarily disallow exporting certain rides that contain unsupported track elements, but thats a topic on its own, its probably also best to to tackle that for once.

This has already been implemented fully since somewhere halfway 2022. (With the addition of new elements for the flying RC, and after that the new elements for the Hybrid, SR and LIM)

> That's why I think something like "Speed Controller" makes more sense where you can just set the right properties and you get pretty much what you want

This is not necessarily needed as a new element as the booster could have regular boost mode as default, you can change the settings to a different mode, if that's selected it can no longer save the track design in current format but that's no biggie. It will also make it a bunch more user friendly than having more elements that basically do the same thing, but one does it better than the other.

**Comment by ZehMatt** [comment]
> This is not necessarily needed as a new element as the booster could have regular boost mode as default, you can change the settings to a different mode, if that's selected it can no longer save the track design in current format but that's no biggie. It will also make it a bunch more user friendly than having more elements that basically do the same thing, but one does it better than the other.

Well Booster to me implies its accelerating, where a Controller is not restricted to just acceleration, the controller would essentially make most things obsolete and it can be probably used as the base element for brakes, boosters and so on where most functionality is just restricted to acceleration or braking. It would be even best if track elements are just objects so we don't have to meddle around in the code anymore as much.

**Comment by karst** [comment]
> Well Booster to me implies its accelerating.

If language is too much of an issue here I think we might just be better off changing the name of the booster to something more fitting. LIMs and LSMs and even Booster tyres can boost and brake a vehicle, so it makes sense having it implemented on the same element as the "booster". Though I feel this discussion would fit better in it's own discussion form than with this specific PR.

**Comment by spacek531** [comment]
The proposed speed controller is out of the scope of this PR. This PR is limited to adding a speed property to block brakes and making adjacent brakes change their speed to match the state of the block brake.

**Comment by ZehMatt** [comment]
> The proposed speed controller is out of the scope of this PR. This PR is limited to adding a speed property to block brakes and making adjacent brakes change their speed to match the state of the block brake.
> 
> Regarding:
> 
> > I'm a bit curious if it wouldn't make perhaps more sense to... ...just have a new element vs hacking new functionality into an existing element?
> 
> Real roller coasters control the brakes on brake runs based on the status of the block brake, which this PR is implementing.

Don't worry not asking to change it here.

**Comment by karst** [comment]
Did you change the behaviour for the brakes yet? And if so have you decided how many units it looks ahead?

**Comment by spacek531** [comment]
I did change the behavior, it looks ahead 256 tiles. I don't think there's much left to do on this one. I'm sure once it's reviewed I will think of something...

File: src/openrct2/ride/Ride.cpp (hunk 16 above)
Comment (spacek531) [reply src/openrct2/ride/Ride.cpp:3058]: this section of code was copied from `block_brakes_open_previous_section`

**Comment by Legotown55** [comment]
I have been testint your speedy blodk breaks and found no problums

**Comment by spacek531** [comment]
I've rebased all the commits so they each should follow a logical order to help reviewing.

File: src/openrct2/ride/Vehicle.cpp [hunk 17]
```diff
@@ -6241,11 +6240,16 @@ static void block_brakes_open_previous_section(
     MapInvalidateElement(location, reinterpret_cast<TileElement*>(trackElement));
 
     auto trackType = trackElement->GetTrackType();
-    if (trackType == TrackElemType::BlockBrakes || trackType == TrackElemType::EndStation)
+    if (ride.IsBlockSectioned())
```
Comment (spacek531) [inline src/openrct2/ride/Vehicle.cpp:6243]: Redundant check - it's called from block_brakes_open_previous_section

**Comment by spacek531** [comment]
Will fix https://github.com/OpenRCT2/OpenRCT2/issues/11649

**Comment by karst** [comment]
That's odd, i thought labelling it as milestone would keep it open.

**Comment by spacek531** [comment]
Please keep open.

**Comment by Legotown55** [comment]
Will this PR ever get reviwed for 0.4.5 plaese

File: src/openrct2/park/ParkFile.cpp [hunk 18]
```diff
@@ -1092,27 +1094,18 @@ namespace OpenRCT2
...
                                     if (TrackTypeMustBeMadeInvisible(
-                                            trackElement->GetRideType(), trackElement->GetTrackType(),
-                                            os.GetHeader().TargetVersion))
+                                            trackElement->GetRideType(), trackType, os.GetHeader().TargetVersion))
                                     {
                                         it.element->SetInvisible(true);
                                     }
-                                }
-                                else if (
-                                    it.element->GetType() == TileElementType::SmallScenery && os.GetHeader().TargetVersion < 22)
```
Comment (duncanspumpkin) [inline src/openrct2/park/ParkFile.cpp:1103]: When you rebase this make sure you don't accidentally delete this code again.

File: src/openrct2/ride/Ride.cpp [hunk 19]
```diff
@@ -3050,20 +3050,75 @@ static void RideOpenBlockBrakes(const CoordsXYE& startElement)
...
             case TrackElemType::DiagUp60ToFlat:
-            case TrackElemType::BlockBrakes:
                 currentElement.element->AsTrack()->SetBrakeClosed(false);
                 break;
         }
     } while (TrackBlockGetNext(&currentElement, &currentElement, nullptr, nullptr)
              && currentElement.element != startElement.element);
 }
 
+void BlockBrakeSetLinkedBrakesClosed(const CoordsXYZ& vehicleTrackLocation, TrackElement& trackElement, bool isClosed)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:3071]: As this is complex and not compeletly intuitive could you add a comment to this function describing what it is doing.

File: src/openrct2/ride/Vehicle.cpp [hunk 20]
```diff
@@ -7376,6 +7377,51 @@ void Vehicle::Sub6DBF3E()
     }
 }
 
+uint8_t Vehicle::ChooseBrakeSpeed() const
+{
+    if (GetTrackType() != TrackElemType::Brakes)
+        return brake_speed;
+    auto trackElement = MapGetTrackElementAtOfTypeSeq(TrackLocation, GetTrackType(), 0);
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Vehicle.cpp:7384]: Also suggest adding a comment to this function as its complex

File: src/openrct2/ride/Vehicle.cpp [hunk 21]
```diff
@@ -7376,6 +7377,51 @@ void Vehicle::Sub6DBF3E()
...
+void Vehicle::GetSpeedFromBrake(const CoordsXYZ& vehicleTrackLocation, TrackElement& brake)
+{
+    auto trackSpeed = brake.GetBrakeBoosterSpeed();
+    brake_speed = trackSpeed;
+    if (brake.GetTrackType() != TrackElemType::Brakes)
+    {
+        BlockBrakeSpeed = trackSpeed;
+        return;
+    }
+    // As soon as feasible, encode block brake speed into track data so the lookforward can be skipped here.
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Vehicle.cpp:7404]: track element data?

File: src/openrct2/ride/Vehicle.cpp [hunk 22]
```diff
@@ -9399,4 +9452,5 @@ void Vehicle::Serialise(DataSerialiser& stream)
     stream << seat_rotation;
     stream << target_seat_rotation;
     stream << BoatLocation;
+    stream << BlockBrakeSpeed;
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Vehicle.cpp:9455]: Add a COMPARE_FILED to gamestatesnapshots.cpp

File: src/openrct2/ride/Track.h (hunk 13 above)
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.h:23]: could you add a leading k to these constants `kBlockBrakeBaseSpeed` its slightly confusing with the new `BlockBrakeSpeed` member variable.

**Comment by karst** [comment]
Out of curiosity which of the functions ended up with this PR? The original braking method or the one that looks ahead?

**Comment by spacek531** [comment]
It uses the bbrake maximum speed which currently requires lookahead

**Review by duncanspumpkin — APPROVED** [review]
